libkbfs/bare_root_metadata_v2.go

```diff
@@ -1136,143 +1061,224 @@ func (md *BareRootMetadataV2) GetCurrentTLFPublicKey(
 	if err != nil {
-		return false, err
+		return nil, err
 	}
-	return ok, nil
-}
 
-// GetUnresolvedParticipants implements the BareRootMetadata interface for BareRootMetadataV2.
-func (md *BareRootMetadataV2) GetUnresolvedParticipants() (readers, writers []keybase1.SocialAssertion) {
-	return md.UnresolvedReaders, md.WriterMetadataV2.Extra.UnresolvedWriters
+	if len(rServerHalves) > 0 {
+		rkb.TLFReaderEphemeralPublicKeys = append(
+			rkb.TLFReaderEphemeralPublicKeys, ePubKey)
+	}
+
+	return rServerHalves, nil
 }
 
-// GetUserDeviceKeyInfoMaps implements the BareRootMetadata interface for BareRootMetadataV2.
-func (md *BareRootMetadataV2) GetUserDeviceKeyInfoMaps(
-	codec kbfscodec.Codec, keyGen KeyGen, _ ExtraMetadata) (
-	readers, writers UserDeviceKeyInfoMap, err error) {
+func (md *BareRootMetadataV2) updateKeyGeneration(
+	crypto cryptoPure, keyGen KeyGen,
+	updatedWriterKeys, updatedReaderKeys UserDevicePublicKeys,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	tlfCryptKey kbfscrypto.TLFCryptKey) (UserDeviceKeyServerHalves, error) {
+	if len(updatedWriterKeys) == 0 {
+		return nil, errors.New(
+			"updatedWriterKeys unexpectedly non-empty")
+	}
+
 	wkb, rkb, err := md.getTLFKeyBundles(keyGen)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	rUDKIM, err := rkb.RKeys.toUDKIM(codec)
+	newIndex := len(wkb.TLFEphemeralPublicKeys)
+
+	wServerHalves, err := wkb.WKeys.fillInUserInfos(
+		crypto, newIndex, updatedWriterKeys, ePrivKey, tlfCryptKey)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	wUDKIM, err := wkb.WKeys.toUDKIM(codec)
+	rServerHalves, err := rkb.RKeys.fillInUserInfos(
+		crypto, newIndex, updatedReaderKeys, ePrivKey, tlfCryptKey)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
 
-	return rUDKIM, wUDKIM, nil
+	serverHalves, err := wServerHalves.mergeUsers(rServerHalves)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(serverHalves) > 0 {
+		wkb.TLFEphemeralPublicKeys =
+			append(wkb.TLFEphemeralPublicKeys, ePubKey)
+	}
+
+	return serverHalves, nil
 }
 
 // AddKeyGeneration implements the MutableBareRootMetadata interface
 // for BareRootMetadataV2.
 func (md *BareRootMetadataV2) AddKeyGeneration(codec kbfscodec.Codec,
-	_ cryptoPure, _ ExtraMetadata,
-	currCryptKey, nextCryptKey kbfscrypto.TLFCryptKey,
-	pubKey kbfscrypto.TLFPublicKey) (ExtraMetadata, error) {
+	crypto cryptoPure, extra ExtraMetadata,
+	updatedWriterKeys, updatedReaderKeys UserDevicePublicKeys,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	pubKey kbfscrypto.TLFPublicKey,
+	currCryptKey, nextCryptKey kbfscrypto.TLFCryptKey) (
+	nextExtra ExtraMetadata,
+	serverHalves UserDeviceKeyServerHalves, err error) {
+	if md.TlfID().IsPublic() {
+		return nil, nil, InvalidPublicTLFOperation{
+			md.TlfID(), "AddKeyGeneration", md.Version()}
+	}
+
+	if len(updatedWriterKeys) == 0 {
```

Comment:
Not sure what you mean -- this is checking that there is at least one writer device, which should always hold when adding a new key generation.

The error message was wrong, though, it should be "unexpectedly empty", not "unexpectedly non-empty". Is that what you meant?